src/main/ruby/core/posix.rb

```diff
@@ -42,6 +42,14 @@ def self.to_nfi_type(type)
     end
   end
 
+  def self.attach_function_lazily(native_name, argument_types, return_type,
+                                  on: Truffle::POSIX, as: native_name,
+                                  library: LIBC, blocking: false)
+    on.define_singleton_method(as) do |*args|
+      attach_function native_name, argument_types, return_type, on: on, as: as, library: library, blocking: blocking
+      send as, *args
+    end
+  end
```

Comment:
Is there any problem to make it lazy by default?
If not then let's make `attach_function` always lazy so we don't need to make the callers know about this (and rename the original attach_function in attach_function_impl/attach_function_eager or so).

Reply:
No issue. However, this is not a user facing function, so why would we hide the fact that it's lazy from ourselves forcing us to remember?

Reply:
Concurrency could be an issue, but it seems it would just define multiple times the same method which is OK here.

It's very clear, the function is defined right above its usages.
The name is longer and these line are already long enough, and it also has the same meaning, no need to leak this implementation detail to each usage.

Reply:
I think making it lazy by default makes sense, providing there aren't any issues. We already make core methods load lazily. The converse argument is I don't want to have to remember which of these to use, when the lazy one should always be used.